---
title: "idpr-vignette"
author: "William McFadden"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{idpr-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6, 
  fig.height = 4
)
```

```{r setup}
library(idpr)
```




~ placeholder for the intro ~ 


## Installation  

The most recent version of the package can be installed with the following line 
of code. This requires the devtools package to be installed.

```{r}
#devtools::install_github("wmm27/idpr")
```

## Profiling

To quickly generate the profile for a protein, a protein sequence and UniProt 
ID is needed. The UniProt ID must be a character string, and the sequence can be
a character string of amino acid residues, a character string specifying the
path to a .fasta file, or amino acid residues as a vector of individual letters.

```{r}
P53_HUMAN <- idpr:::TP53Sequences[2] #Getting a preloaded sequence from idpr
print(P53_HUMAN)

P53_ID <- "P04637" #Human TP53 UniProt ID
```

Then we can generate the profile with idprofile()

```{r}
idprofile(sequence = P53_HUMAN,
          uniprotAccession = P53_ID)
```


idprofile returns 4-5 plots:

 * Charge-Hydropathy Plot^\*^
 * Plot of Amino Acid Composition and Structural Tendency^†^
 * Calculations of Local Charge Along a Protein Sequence^\*^
 * Local, Scaled Hydropathy Along a Protein Sequence^\*^
 * A prediction of intrinsic disorder by IUPred2 (only with a uniprotAccession)^‡^
 
*Detailed descriptions of each plot can be found in specific vignettes.*

\* *Charge Hydropathy Vignette*

† *Structural Tendency Vignette*

‡ *IUPred Vignette; see Erdős, G., & Dosztányi, Z. (2020).*


A brief explanation of each plot is given below:


### Charge-Hydropathy Plot

It was shown by Uversky, Gillespie, & Fink (2000) that both high net charge and 
low mean hydropathy are properties of IDPs. One explanation is that a high net 
charge leads to increased repulsion of residues causing an extended structure 
and low hydrophobicity will reduce the hydrophobic interactions causing reduced 
protein packing. When both average net charge and mean scaled hydropathy are 
plotted, IDPs occupy a unique area on the plot. Therefore, this can be used to 
distinguish proteins that are extended or compact under native conditions. 
However, it is important to note that IDPs can have the characteristics of
a collapsed protein or an extended protein (1, 2). 

### Structural Tendency Plot

The composition of amino acids and the overall chemistry of Intrinsically
Disordered Proteins (IDPs) are distinctly different from that of ordered 
proteins. Each amino acid has a tendency to favor a compact or extended 
secondary and tertiary structures based on the chemistry of the residue.
Disorder-promoting residues, those enriched in IDPs, are typically hydrophilic, 
charged, or small. Order promoting residues, those enriched in 
structured proteins, tend to be aliphatic, hydrophobic, aromatic, or form
tertiary structures. Disorder neutral residues neither favor order nor disordered
structures (3).  
Disorder-promoting residues are P, E, S, Q, K, A, and G; 
order-promoting residues are M, N, V, H, L, F, Y, I, W, and C; 
disorder‐neutral residues are D, T, and R (4). 


### Local Charge Calculations

As stated, IDPs are enriched in charged residues. Residues of similar charge 
tend to repel each other which can prevent protein packing and help lead to an 
unstructured protein under native conditions (1). There are many pKa data sets, 
so we utilize the IPC pKa data set for calculations (5). Identifying local 
charge is an important biochemical measurement beyond the scope intrinsic 
disorder. Charges are calculated using a sliding window to help identify regions 
of extreme charge. The resulting figure is similar to ProtScale from ExPASy (6).

This package is preloaded with several pKa data sets and allows for custom
pKa sets to be used. See the Charge-Hydropathy Vignette for additional details.

### Local Hydropathy

As stated, IDPs are depleted in hydrophobic residues (1). The hydrophobic effect 
is a significant driving force in protein packing and leads rigid structures 
(7). IDPs lack this driving force and have residues that preferentially interact
with the solvent (3). This uses the Kyte and Doolittle measurement of hydropathy 
(8), scaled with Arg having a hydropathy of 0 and Ile having a hydropathy of 1.
This was done to parallel the work in the Charge-Hydropathy plot which utilizes
the same, normalized scale (1, 3).  The resulting figure is similar to ProtScale
from ExPASy (6).
Identifying hydropathy is an important biochemical measurement beyond the 
scope intrinsic disorder. Scaled hydropathy is averaged locally along the
protein using a sliding window to identify regions devoid of hydropathic 
characteristics.


### IUPred

IUPred2 analyzes an amino acid sequence and returns a score of 
intrinsic disorder depending on a model of the estimated energy potential for 
residue interactions (9) This is because 
structured proteins have the ability to create a network of interactions, while
IDPs lack abundant interactions. The reduced number of interactions leads to an 
IDP's lack of secondary and tertiary structure (10). 
Predictions are made on a scale of 0-1, where any residues with a score over 
0.5 are predicted to be in a disordered region, and any residue scoring below 
0.5 are predicted to be ordered (9-11).

The IUPred graph shown is the default setting, which is the prediction of
long disordered regions. IUPred2A Offers several predictors. An additional
prediction of protein-protein interactions is done with the ANCHOR2 method, 
and another predictor of redox-sensitive disorder prediction is done with 
IUPred2A Redox (9-11).

These plots can be generated independently, shown here, or with idprofile()
using the iupredType argument. 

```{r}
iupredAnchor(P53_ID) #IUPred2 long + ANCHOR2 prediction of scaffolding
```

Redox-sensitive regions are shaded with a green background.
```{r}
iupredRedox(P53_ID) #IUPred2 long with enviornmental context
```

***

## Visualizing Discrete Values

As mentioned in the “Structural Tendency Plot” section above, there are specific amino acid residues that are enriched in unstructured protein regions and other amino acids that favor ordered protein regions (4). While the total composition is important to know, the location of these are important to visualize especially in proteins with both ordered and disordered domains. 

Continuous values, like charge and disorder predictions, are easy to visualize along a protein sequence, while discrete values can be more challenging. ‘idpr’ contains a way to visualize discrete values of a protein sequence with the sequenceMap() function. The workflow shown here is how to retrieve a data frame of structural tendency of each residue in a sequence, followed by visualizing.


```{r}
p53_tendency_DF <- structuralTendency(P53_HUMAN)
head(p53_tendency_DF) #see the first few rows of the generated data frame

sequenceMap(sequence = P53_HUMAN,
            property = p53_tendency_DF$Tendency)  #generate the map
```


sequenceMap() does accept continuous values as well. Additionally, custom plots
that match the theme of plots generated by idprofile() and other **idpr** 
functions can be created using the sequencePlot() function within **idpr**

***


## Substitution Matrices for Analyzing IDPs

In addition to differing biochemistry between IDPs and ordered proteins, IDPs
tend to experience evolution differently. IDPs and IDRs tend to evolve at a
faster rate than ordered proteins or folded domains (12, 13). Additionally, IDPs
tend to accept different point mutations when compared to ordered proteins (14). 

Currently, the most commonly used amino acid substitution matrices are PAM and 
BLOSUM (15, 16). These are integrated into many web-based tools like those on 
EMBOSS and NCBI-BLAST+ (17, 18). These do not allow for custom matrices. BLOSUM
and PAM matrices are data that can be loaded with the Biostrings Package and can 
be used with alignment programs in R (19). 

Something important to note when using these matrices, is that both PAM and 
BLOSUM are derived from or favor structured proteins and therefore are not the 
most appropriate to use when analyzing IDPs (14, 20, 21). Trivedi & Nagarajaram
(2019) provides comparisons between commonly used substitution matrices and
matrices developed for analyzing IDPs. They additionally developed a matrix that
is better at identifying homologs of IDPs (20). 

Three groups of IDP-derived substitution matrices are incorporated into the
idpr package for use in alignments done with R. These are the matrices from 
Trivedi & Nagarajaram (2019) (20), Brown et al. (2009) (14), and Radivojac 
et al. (2001) (21). 




## References

1.    Uversky VN, Gillespie JR, Fink AL. Why are “natively unfolded” proteins unstructured under physiologic conditions? Proteins: Structure, Function, and Bioinformatics. 2000;41(3):415-27.
2.    Uversky VN. Paradoxes and wonders of intrinsic disorder: Complexity of simplicity. Intrinsically Disordered Proteins. 2016;4(1):e1135015.
3.    Uversky VN. Intrinsically Disordered Proteins and Their “Mysterious” (Meta)Physics. Frontiers in Physics. 2019;7(10).
4.    Uversky VN. Unusual biophysics of intrinsically disordered proteins. Biochimica et Biophysica Acta (BBA) - Proteins and Proteomics. 2013;1834(5):932-51.
5.    Kozlowski LP. IPC – Isoelectric Point Calculator. Biology Direct. 2016;11(1):55.
6.    Gasteiger E, Hoogland C, Gattiker A, Wilkins MR, Appel RD, Bairoch A. Protein identification and analysis tools on the ExPASy server.  The proteomics protocols handbook: Springer; 2005. p. 571-607.
7.    Durell SR, Ben-Naim A. Hydrophobic-hydrophilic forces in protein folding. Biopolymers. 2017;107(8):10.1002/bip.23020.
8.    Kyte J, Doolittle RF. A simple method for displaying the hydropathic character of a protein. Journal of molecular biology. 1982;157(1):105-32.
9.    Mészáros B, Erdős G, Dosztányi Z. IUPred2A: context-dependent prediction of protein disorder as a function of redox state and protein binding. Nucleic acids research. 2018;46(W1):W329-W37.
10.    Dosztányi Z. Prediction of protein disorder based on IUPred. Protein Sci. 2018;27(1):331-40.
11.    Erdős G, Dosztányi Z. Analyzing Protein Disorder with IUPred2A. Current Protocols in Bioinformatics. 2020;70(1):e99.
12.    Franzosa EA, Xia Y. Structural Determinants of Protein Evolution Are Context-Sensitive at the Residue Level. Molecular Biology and Evolution. 2009;26(10):2387-95.
13.    Brown CJ, Takayama S, Campen AM, Vise P, Marshall TW, Oldfield CJ, et al. Evolutionary rate heterogeneity in proteins with long disordered regions. Journal of molecular evolution. 2002;55(1):104.
14.    Brown CJ, Johnson AK, Daughdrill GW. Comparing Models of Evolution for Ordered and Disordered Proteins. Molecular Biology and Evolution. 2009;27(3):609-21.
15.    Dayhoff M, Schwartz R, Orcutt B. 22 a model of evolutionary change in proteins.  Atlas of protein sequence and structure. 5: National Biomedical Research Foundation Silver Spring MD; 1978. p. 345-52.
16.    Henikoff S, Henikoff JG. Amino acid substitution matrices from protein blocks. Proc Natl Acad Sci U S A. 1992;89(22):10915-9.
17.    Johnson M, Zaretskaya I, Raytselis Y, Merezhuk Y, McGinnis S, Madden TL. NCBI BLAST: a better web interface. Nucleic Acids Research. 2008;36(suppl_2):W5-W9.
18.    Madeira F, Park YM, Lee J, Buso N, Gur T, Madhusoodanan N, et al. The EMBL-EBI search and sequence analysis tools APIs in 2019. Nucleic acids research. 2019;47(W1):W636-W41.
19.    Pagès H, Aboyoun P, Gentleman R, DebRoy S. Biostrings: Efficient manipulation of biological strings. R package version. 2020;2(0).
20.    Trivedi R, Nagarajaram HA. Amino acid substitution scoring matrices specific to intrinsically disordered regions in proteins. Scientific Reports. 2019;9(1):16380.
21.    Radivojac P, Obradovic Z, Brown CJ, Dunker AK. Improving sequence alignments for intrinsically disordered proteins.  Biocomputing 2002: World Scientific; 2001. p. 589-600.





